### PR TITLE
Grid generation: multiple nest capability

### PR DESCRIPTION
**Pull request by A-Kyle**
This branch includes Bill Ramstrom's (AOML/HRD) original modifications to the make_hgrid code that allow it to generate multiple nested grids on the cubed sphere. The usage message for the program has been updated to reflect changes to its functionality. 

Example command-line statement to generate grids for all 6 tiles of the cubed sphere and 2 nests on tile 6:

`./make_hgrid --grid_type gnomonic_ed --nlon 192 --grid_name C96_grid --do_schmidt --stretch_factor 1. --target_lon -57.0 --target_lat 0.0 --nest_grids 2 --parent_tile 6,6 --refine_ratio 4,4 --istart_nest 81,151 --jstart_nest 81,39 --iend_nest 112,182 --jend_nest 112,70 --halo 3 --great_circle_algorithm`

### CODE REVIEW

**Comment by bensonr** [comment]
@ceblanton @ngs333 

Can you help this user get things working as expected - thanks


ADDING ORIGINAL CORRESPONDENCE FROM EMAIL:

You can find the most recent logs at https://github.com/NOAA-GFDL/FRE-NCtools/pull/27/checks?check_run_id=1513478791 ; I've also attached a .zip with the logs in them. The problem in question appears while running Test04-grid_coupled_nest.sh , when trying to run make_hgrid with the arguments I mentioned earlier. Again, after finding a bug earlier and applying a simple fix, I was able to run make_hgrid with the arguments myself, but the workflow online still seems to run into an issue there.

-Kyle

On Mon, Dec 7, 2020 at 6:10 PM Kyle Ahern - NOAA Affiliate <kyle.ahern@noaa.gov> wrote:
All,

As a follow-up, I wanted to note that I found a bug that was causing a segmentation fault when running make_hgrid with global refinement (parent_tile=0). Locally, I am now able to run the command I mentioned earlier, but there still seems to be an issue arising in the test workflow on github that I can't identify (non-zero return status that I can't replicate). I'm not sure where to go from there, so let me know if you have recommendations for moving further on this.

 -Kyle

On Mon, Dec 7, 2020 at 12:37 PM Kyle Ahern - NOAA Affiliate <kyle.ahern@noaa.gov> wrote:
Hi all,

After the FV3 Team Meeting last Friday, I cloned the NOAA-GFDL/FRE-NCtools git repo that Rusty provided and applied the edits I currently have based on Bill Ramstrom's original modifications to the make_hgrid code to implement multiple nest grid generation. After a pull request, the test workflow ran into two issues: one to do with hydrology (which seems to be a problem with the c-shell interpreter unrelated to these code changes), and the other to do with creating a C144 grid.

Specifically, the command that failed was:
#Create C144 grid for land
run command make_hgrid  \
    --grid_type gnomonic_ed  \
    --nlon 288  \
    --grid_name land_grid  \
    --do_schmidt \
    --target_lat 48.15  \
    --target_lon -100.15  \
    --halo 3  \
    --stretch_factor 3  \
    --great_circle_algorithm  \
    --nest_grid  \
    --refine_ratio 3  \
    --parent_tile 0

I'll look around to see if I can pin down the problem today. My suspicion is that it has something to do with parent_tile being set to 0. You can see the test workflow's logs at https://github.com/NOAA-GFDL/FRE-NCtools/runs/1501435813 . I'm green to these kinds of workflows and collaborative testing procedures, so please let me know if I've misinterpreted these errors or processes, or if I should go about PRs for these projects differently.

Thanks,
-Kyle

**Comment by ngs333** [comment]
I agree that the hydrology issue is not related to the code changes.

Also, I cloned Kyles fork and branch, rand on gaea with Intel, and test # 4
passed. I can't determine from the log what happened with the CI failure on
this test,
but I will see later if I can run this on a different platform with gcc.

**Comment by A-Kyle** [comment]
Hi Miguel,

I doubt this is the case, but is it possible that the print statements added to make_hgrid.c and/or create_gnomonic_cubic_grid.c could be triggering a non-zero return code? I noticed that the log included a line after test 04 failed, "string: 3", which is a line typically printed in the new code. Either way, in seeing that line, I realize I should probably disable those statements for non-debug operations.

-Kyle

**Comment by ceblanton** [comment]
Hello, I did as Miguel suggested- checking out Kyle branch and fork and compiling with gcc/gfortran (on my home Ubuntu), and Test 4 passed. I'll try running the commands interactively, but we might need to expand the github CI logs to see what is causing the failure.

(Also, completely agree that the hydrology script test failure is unrelated.)

File: tools/make_hgrid/make_hgrid.c [hunk 1]
```diff
@@ -375,7 +426,7 @@ void fill_cubic_grid_halo(int nx, int ny, int halo, double *data, double *data1_
     }
 
     for(i=1; i<=nxp; i++) {
-      data[i] = data2_all[ls*nxp*nyp+(nxp-i)*nyp+(nx-1)]; /*south */
+      data[i] = data2_all[ls*nxp*nyp+(nxp-i)*nxp+(nx-1)]; /*south */
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:429]: Was this, and the other two on line 440 and 445 errors in the original version?

File: tools/make_hgrid/make_hgrid.c [hunk 2]
```diff
@@ -891,38 +981,42 @@ int main(int argc, char* argv[])
...
-    
-    size1 = (unsigned long)nxp*nyp*ntiles;
-    size2 = (unsigned long)nxp*(nyp+1)*ntiles;
-    size3 = (unsigned long)(nxp+1)*nyp*ntiles;
-    size4 = (unsigned long)nxp    *nyp*ntiles;
+    int size1, size2, size3, size4;
+    int n_nest;
+
+    for (n_nest=0; n_nest < ntiles; n_nest++) {
+      printf("[INFO] tile: %d, nxl[%d], nyl[%d], ntiles: %d\n", n_nest, nxl[n_nest], nyl[n_nest], ntiles);
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:991]: Shouldn't this be printed to stderr?

File: tools/make_hgrid/make_hgrid.c [hunk 3]
```diff
@@ -891,38 +981,42 @@ int main(int argc, char* argv[])
...
+      if (verbose) printf("[INFO] Adding memory size for nest %d, nest_grids: %d\n", n_nest, nest_grids);
+      size1 += (nxl[n_nest] +1)    *  (nyl[n_nest]+1);
+      size2 += (nxl[n_nest] +1)    * (nyl[n_nest]+2);
+      size3 += (nxl[n_nest]+2) *  (nyl[n_nest]+1);
+      size4 += (nxl[n_nest]+1)     *  (nyl[n_nest]+1);
     }
 
+
+
+    if (verbose) printf("[INFO] Allocating arrays of size %d for x, y based on nxp: %d nyp: %d ntiles: %d\n", size1, nxp, nyp, ntiles);
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:1011]: stderr?

File: tools/make_hgrid/make_hgrid.c [hunk 4]
```diff
@@ -940,7 +1034,7 @@ int main(int argc, char* argv[])
 			 area, angle_dx, center, verbose, use_great_circle_algorithm);
   else if(my_grid_type==FROM_FILE) {
     for(n=0; n<ntiles; n++) {
-      long n1, n2, n3, n4;
+      int n1, n2, n3, n4;
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:1037]: Why the change from `long` to `int`?

File: tools/make_hgrid/make_hgrid.c [hunk 5]
```diff
@@ -984,10 +1080,16 @@ int main(int argc, char* argv[])
...
+
+      if (verbose) {
+	printf("Writing out %s.\n", outfile);
+      }
+
       fid = mpp_open(outfile, MPP_WRITE);
       /* define dimenison */
       nx = nxl[n];
       ny = nyl[n];
+      if (verbose) printf("[INFO] Outputting arrays of size nx: %d and ny: %d for tile: %d\n", nx, ny, n);
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:1092]: stderr?

File: tools/make_hgrid/make_hgrid.c [hunk 6]
```diff
@@ -1057,63 +1155,79 @@ int main(int argc, char* argv[])
       mpp_put_var_value_block(fid, id_tile, start, nwrite, tilename );
 
       if(out_halo ==0) {
+	if (verbose) printf("[INFO] START NC XARRAY write out_halo=0 tile number = n: %d offset = pos_c: %d\n", n, pos_c);
+	if (verbose) printf("[INFO] XARRAY: n: %d x[0]: %f x[1]: %f x[2]: %f x[3]: %f x[4]: %f x[5]: %f x[10]: %f\n", n, x[pos_c], x[pos_c+1], x[pos_c+2], x[pos_c+3], x[pos_c+4], x[pos_c+5], x[pos_c+10]);
+	if (verbose && n > 0) printf("[INFO] XARRAY: n: %d x[0]: %f x[-1]: %f x[-2]: %f x[-3]: %f x[-4]: %f x[-5]: %f x[-10]: %f\n", n, x[pos_c], x[pos_c-1], x[pos_c-2], x[pos_c-3], x[pos_c-4], x[pos_c-5], x[pos_c-10]);
```
Comment (underwoo) [inline tools/make_hgrid/make_hgrid.c:1160]: stderr?  Also, why not just one `if (verbose)` block, with a nested `if (n > 0)`?

Note, there are other `printf("[INFO] ...` that should be corrected.  I'm not going to continue marking them.  I  feel diagnostics, notices, warnings, etc. should be sent to stderr.

**Review by underwoo — CHANGES_REQUESTED** [review]

**Comment by underwoo** [comment]
> (Also, completely agree that the hydrology script test failure is unrelated.)

The hydrology failure is unrelated, and should now be fixed via #22.

File: tools/make_hgrid/make_hgrid.c (hunk 1 above)
Comment (A-Kyle) [reply tools/make_hgrid/make_hgrid.c:429]: Thanks for pointing this out. After investigating commit histories here and in the HAFS-community git repo (which is where I originally obtained the code for make_hgrid), my guess is that the changes I have are probably outdated.

**Explanation:** 
It looks like the version I have is similar to what was in the NOAA-GFDL/FRE-NCtools master branch _prior_ to edits in September 2015 by Zhi Liang, which _appear_ to be bug fixes (see lines 363, 374, 379 at https://github.com/NOAA-GFDL/FRE-NCtools/commit/168f6a5a2717bc80961da6ba751509cd8c7ace0d ). These lines have remained the same at NOAA-GFDL's master branch up until now. However, the changes that Bill Ramstrom and I have made to the source relative to what was in the HAFS-community git repo did not include differences at these lines (363, 374, 379 at https://github.com/hafs-community/UFS_UTILS/commit/11b3a182a5a5407e7005a6b20f56473b8c524f99 ), which indicates to me that the code in the HAFS repo did not include Zhi Liang's changes in 2015 and thus may be out of date. Looking at the initial commit for make_hgrid.c in the HAFS repo, it looks like these lines were never changed there.

I'll modify the code I have to reflect the prior version and submit a new commit, along with changes to the print statements.

-Kyle

File: tools/make_hgrid/make_hgrid.c (hunk 4 above)
Comment (A-Kyle) [reply tools/make_hgrid/make_hgrid.c]: Changed this back to the more-specific `long` type.

File: tools/make_hgrid/make_hgrid.c (hunk 6 above)
Comment (A-Kyle) [reply tools/make_hgrid/make_hgrid.c]: I've gone back and adjusted these `printf("[INFO] ...` statements to be piped to stderr. I've also condensed consecutive `if (verbose)` statements into blocks.

**Comment by ngs333** [comment]
Kyle,
I tried merging your fork/branch into master  within a pre-merge testing area but the automatic merge could not finish. All the conflicts are in a single file : create_gnomonic_cubic_grid.c. I believe your version should be the final or near final version at this point, and also that the auto merge failed do to an earlier merge I made having to do with moving to C99 standards compliance.
In the process of doing the merge work, I made this branch on my fork : https://github.com/ngs333/FRE-NCtools/tree/multinests_a.  It has the latest master (which is from NOAA-GFDL/FRE-NCTools) plus your merge.  In resolving the merge conflicts and removing the compilation warning desired since C99 compliance, I made  changes:
a) Changing "};"  to "}" at the end of function definitions (there were about a dozen of those)
b) Changing "%d" to percent "%ld" in several sprintf statements.
Additionally, there were two compiler warnings related to printing that I think we should take care of perhaps and that you would be in a better position to decide what to do with them. These warning are:
```c 
create_gnomonic_cubic_grid.c:1661:105: warning: format ‘%p’ expects argument of type ‘void *’, but argument 6 has type ‘const double *’ [-Wformat=]
 1661 |    fprintf(stderr, "[INFO] setup_aligned nest: parent_ni: %d parent_nj: %d refine_ratio: %d parent_xc: %p parent_yc: %p\n",
      |                                                                                                        ~^
      |                                                                                                         |
      |                                                                                                         void *
 1662 |             parent_ni, parent_nj, refine_ratio, parent_xc, parent_yc);
      |                                                 ~~~~~~~~~                                                
      |                                                 |
      |                                                 const double *
make_hgrid/create_gnomonic_cubic_grid.c:1661:119: warning: format ‘%p’ expects argument of type ‘void *’, but argument 7 has type ‘const double *’ [-Wformat=]
 1661 |    fprintf(stderr, "[INFO] setup_aligned nest: parent_ni: %d parent_nj: %d refine_ratio: %d parent_xc: %p parent_yc: %p\n",
      |                                                                                                                      ~^
      |                                                                                                                       |
      |                                                                                                                       void *
 1662 |             parent_ni, parent_nj, refine_ratio, parent_xc, parent_yc);
      |                                                            ~~~~~~~~~                                                   
      |                                                            |
      |                                                            const double *
```

I think the base way to proceed if for you to either a) merge the current master into  your form and branch, or b) merge my fork and branch into yours. At this point I thing b) would be the easiest

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 7]
```diff
@@ -94,51 +139,103 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+  for (nn=0; nn < MAX_NESTS; nn++) {
+    ni_nest[nn] = 0;
+    nj_nest[nn] = 0;  
+  }
+
   ntiles2=ntiles;
   global_nest=0;
-  if(nest_grid && parent_tile== 0)
+
+  if (verbose) fprintf(stderr, "[INFO] B\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:151]: This `fprintf` appears to be a leftover debug statement.  If you want a info statement here, it should be than just "B", or just remove this line.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 8]
```diff
@@ -148,6 +245,8 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
   lon = (double *)malloc(nip*nip*sizeof(double));
   lat = (double *)malloc(nip*nip*sizeof(double));
   
+  if (verbose) fprintf(stderr,"[INFO] E\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:248]: More leftover debug print statements.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 9]
```diff
@@ -159,10 +258,35 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
 
   symm_ed(ni, lon, lat);
 
+  if (verbose) fprintf(stderr,"[INFO] F\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:261]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 10]
```diff
@@ -177,13 +301,18 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
      and greenwich meridian Lon[-pi,pi]  */
   mirror_grid(ni, ntiles, xc, yc);
 
+  if (verbose) fprintf(stderr, "[INFO] G\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:304]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 11]
```diff
@@ -177,13 +301,18 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
   for(n=0; n<ntiles*nip*nip; n++) {
     /* This will result in the corner close to east coast of china */
-    if( do_schmidt == 0 && do_cube_transform==0 && shift_fac > EPSLN4) xc[n] -= M_PI/18.;
+    if( do_schmidt == 0 && do_cube_transform == 0 && shift_fac > EPSLN4) xc[n] -= M_PI/18.;
     if(xc[n] < 0.) xc[n] += 2.*M_PI;
     if(fabs(xc[n]) < EPSLN10) xc[n] = 0;
     if(fabs(yc[n]) < EPSLN10) yc[n] = 0;
   }
+
+  if (verbose) fprintf(stderr, "[INFO] H\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:315]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 12]
```diff
@@ -221,20 +350,30 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
     yc[5*nip*nip+j*nip] = yc[4*nip*nip+j*nip+ni];    /* 5E -> 6W */  
   }
 
+  if (verbose) fprintf(stderr, "[INFO] I\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:353]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 13]
```diff
@@ -221,20 +350,30 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+    for (n=0; n<ntiles; n++) {
+
+      if (verbose) fprintf(stderr, "[INFO] Calling cube_transform for tile %d\n", n);
+
+      cube_transform(stretch_factor, 0, ni, 0, ni, target_lon*D2R, target_lat*D2R,
+           n, xc+n*nip*nip, yc+n*nip*nip);
     }
   }
 
+  if (verbose) fprintf(stderr, "[INFO] J\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:375]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 14]
```diff
@@ -279,118 +437,195 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
   nj2p = nj2+1;
   xtmp = (double *)malloc(ni2p*nj2p*sizeof(double));
   ytmp = (double *)malloc(ni2p*nj2p*sizeof(double));
-  
+
+  if (verbose) fprintf(stderr, "[INFO] L\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:441]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 15]
```diff
@@ -279,118 +437,195 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+	  if (n1 < min_n1) min_n1 = n1;
+	  if (n1 > max_n1) max_n1 = n1;
+	}
+
+
+      }
 
     /* cell north and copy to super grid */
-    cell_north(nil[n], njl[n], xc+n*nip*nip, yc+n*nip*nip, xtmp, ytmp);
+    //cell_north(nil[n], njl[n], xc+n*nip*nip, yc+n*nip*nip, xtmp, ytmp);
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:526]: Why is the old line left in as a comment?

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 16]
```diff
@@ -279,118 +437,195 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+    cell_north(nil[n], njl[n], xc + tile_offset[n], yc + tile_offset[n], xtmp, ytmp);
     for(j=0; j<=njl[n]; j++) for(i=0; i<nil[n]; i++) {
-      n1 = n*nxp*nxp+(j*2)*(2*nil[n]+1)+i*2+1;
-      n2 = j*nil[n]+i;
-      x[n1]=xtmp[n2];
-      y[n1]=ytmp[n2];
-    }
-  } 
+	// Offset of 1 for i=0, j=0
+	//n1 = n*nxp*nxp+(j*2)*(2*nil[n]+1)+i*2+1;
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:530]: Here as well, old line is a comment.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 17]
```diff
@@ -279,118 +437,195 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+
+    if (verbose) fprintf(stderr,
+                         "[INFO] INDEX tile: %d min_n1: %d max_n1: %d max_n1 - min_n1: %d sqrt(max_n1 - min_n1 + 1): %f\n",
+                         n, min_n1, max_n1, max_n1 - min_n1, sqrt(max_n1 - min_n1 + 1));
+  }
 
   free(xtmp);
   free(ytmp);
+
+  if (verbose) fprintf(stderr, "[INFO] M\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:551]: Another random `fprintf`.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 18]
```diff
@@ -399,81 +634,78 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
     }
+    calc_cell_area(nx_nest_arr[nn], ny_nest_arr[nn],
+                   x + tile_offset_supergrid[ntiles+nn],
+                   y + tile_offset_supergrid[ntiles+nn],
+                   area + tile_offset_area[ntiles+nn]);
   }
-  
-  if(output_length_angle) {
+
+  if (verbose) fprintf(stderr, "[INFO] Q\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:678]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 19]
```diff
@@ -399,81 +634,78 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
-    /* convert grid location from radians to degree */
-    npts = ntiles*nxp*nyp;
-    if(nx_nest>0) npts += (nx_nest+1)*(ny_nest+1);
 
-  
-    for(i=0; i<npts; i++) {
-      x[i] = x[i]*R2D;
-      y[i] = y[i]*R2D;
-    }
+  if (verbose) fprintf(stderr, "[INFO] R\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:693]: Another.

File: tools/make_hgrid/create_gnomonic_cubic_grid.c [hunk 20]
```diff
@@ -399,81 +634,78 @@ void create_gnomonic_cubic_grid( char* grid_type, int *nlon, int *nlat, double *
...
+    x[i] = x[i]*R2D;
+    y[i] = y[i]*R2D;
   }
+
   free(xc);
-  free(yc);  
-  
+  free(yc);
+
+  if (verbose) fprintf(stderr, "[INFO] Z\n");
```
Comment (underwoo) [inline tools/make_hgrid/create_gnomonic_cubic_grid.c:707]: Another.

**Review by underwoo — CHANGES_REQUESTED** [review]

**Review by ngs333 — CHANGES_REQUESTED** [review]
My comments in the conversation on 12/11 about 5PM ff.

**Comment by A-Kyle** [comment]
@ngs333 Hey Mike,

I've merged in your branch as suggested. As for the problem parameters you noted in the fprintf statements, I typecasted those two arguments to become void pointers, which should take care of the compiler warnings:
`fprintf(stderr, "[INFO] setup_aligned nest: parent_ni: %d parent_nj: %d refine_ratio: %d parent_xc: %p parent_yc: %p\n",`
`parent_ni, parent_nj, refine_ratio, (void *)parent_xc, (void *)parent_yc);`

-Kyle

**Comment by A-Kyle** [comment]
@underwoo Hey Seth,

I've gone back and removed these excess print statements, as well as the old lines of code that were commented out in create_gnomonic_cubic_grid.c . I also cut a couple of print lines that were redundant.

-Kyle

File: tools/make_hgrid/create_gnomonic_cubic_grid.c (hunk 7 above)
Comment (A-Kyle) [reply tools/make_hgrid/create_gnomonic_cubic_grid.c]: I think I've removed all these short print statements in the latest commit. I also took out the old commented-out code that you mentioned in a comment below.

**Review by ngs333 — APPROVED** [review]

**Comment by ngs333** [comment]
As far as I can tell, all the requested changes have been made so I should merge this PR soon.

**Comment by ngs333** [comment]
@A-Kyle Its not required for this PR,  but I am going to write an issue for contributing a "unit test" for this code. Do you also have a test you wish to contribute?

**Review by bensonr — COMMENTED** [review]
I would like to see an example use case added at the bottom of the usage block detailing the command-line for generating nests and/or multiple nests within the new structure.

**Comment by A-Kyle** [comment]
@ngs333 @bensonr I've added a multi-nest example to the list in the usage statement in make_hgrid.c . A test that I used to run the multiple nest capability (along with the mosaic step) used these options (styled after the Test-04 script):

`# Create C96 grid with two nests on tile 6. `
`run command make_hgrid  \ `
`    --grid_type gnomonic_ed  \`
`    --nlon 192  \`
`    --grid_name nested_grid  \`
`    --do_schmidt  \`
`    --target_lat 23.0  \`
`    --target_lon -47.0  \`
`    --halo 3  \`
`    --stretch_factor 3  \`
`    --great_circle_algorithm  \`
`    --nest_grids 2  \`
`    --refine_ratio 2,2  \`
`    --parent_tile 6,6  \`
`    --istart_nest 11,51  \`
`    --jstart_nest 11,51  \`
`    --iend_nest 42,82  \`
`    --jend_nest 42,82`
`  [ "$status" -eq 0 ]`

`# Create C96 mosaic with two nests on tile 6.`
`run command make_solo_mosaic  \`
`    --num_tiles 8  \`
`    --dir ./  \`
`    --mosaic nested_mosaic  \`
`    --tile_file nested_grid.tile1.nc,nested_grid.tile2.nc,nested_grid.tile3.nc,nested_grid.tile4.nc,nested_grid.tile5.nc,nested_grid.tile6.nc,nested_grid.tile7.nc,nested_grid.tile8.nc `
`  [ "$status" -eq 0 ]`

**Review by bensonr — APPROVED** [review]
